src/ParallelTempering.jl

```diff
@@ -0,0 +1,168 @@
+using MPI
+
+""" 
+Container for sampler type as there is no deriving from concrete types in Julia
+"""
+mutable struct Replica
+    rank::Int64
+    N_ranks::Int64
+    rex_dir::Int64
+    nn_ranks::Vector{Int64}
+    nn_βs::Vector{Float64}
+
+    sampler::AbstractSampler
+end
+
+
+"""
+Constructor which initializes MPI communicator and sets sampler.
+-> Use on single communicator with no groups for now
+"""
+function Replica(s::AbstractSampler)
+    # initialize MPI communicator and variables
+    MPI.Init()
+    MPI_COMM_WORLD = MPI.COMM_WORLD
+
+    rank = MPI.Comm_rank(MPI_COMM_WORLD)
+    N_ranks = MPI.Comm_size(MPI_COMM_WORLD)
+
+    Random.seed!(round(Int64, time()*1000))
+
+    # even rank exch. down when rex_dir==0, up when rex_dir==1
+    nn_ranks = [
+        rank + 2*(rank%2) - 1,
+        rank + 1 - 2*(rank%2)
+    ]
```

Comment:
This may just be me, but I think it's a bit more interpretable if `nn_ranks` is your left/right neighbor for all ranks, and then you start off even ranks with `rex_dir = 1` and odd ranks with `rex_dir=2`.

Then, in `run!` you flip with `rex_dir = 3 - rex_dir` and you no longer need `rex_ID` in `replica_exchange!`. Might as well stay in the 1-indexing.

Reply:
This seems like a good idea.

Reply:
Thanks for the nice suggestion - this makes a lot more sense! I've changed it to this now